src/messaging/ExchangeContext.cpp

```diff
@@ -308,25 +380,73 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
     // layer has completed its work on the ExchangeContext.
     Retain();
 
+    messageId   = packetHeader.GetMessageId();
     protocolId  = payloadHeader.GetProtocolID();
     messageType = payloadHeader.GetMessageType();
 
-    // Since we got the response, cancel the response timer.
-    CancelResponseTimer();
+    if (payloadHeader.IsAckMsg())
+    {
+        err = mReliableMessageContext.HandleRcvdAck(payloadHeader.GetAckId().Value());
+    }
 
-    // If the context was expecting a response to a previously sent message, this message
-    // is implicitly that response.
-    SetResponseExpected(false);
+    if (payloadHeader.IsNeedsAck())
+    {
+        MessageFlags msgFlags;
 
-    if (mDelegate != nullptr)
+        // Set the flag in message header indicating an ack requested by peer;
+        msgFlags.Set(MessageFlagValues::kMessageFlag_PeerRequestedAck);
+
+        // Set the flag in the exchange context indicating an ack requested;
+        mReliableMessageContext.SetPeerRequestedAck(true);
+
+        if (!mReliableMessageContext.ShouldDropAckDebug())
+        {
+            err = mReliableMessageContext.HandleNeedsAck(messageId, msgFlags);
+        }
+    }
+
+    // Received Flow Throttle
+    if (protocolId == Protocols::kProtocol_Protocol_Common && messageType == Protocols::Common::kMsgType_RMP_Throttle_Flow)
     {
-        mDelegate->OnMessageReceived(this, packetHeader, protocolId, messageType, std::move(msgBuf));
+        const uint8_t * p        = nullptr;
+        uint32_t PauseTimeMillis = 0;
+
+        // Extract PauseTimeMillis from msgBuf
+        p                  = msgBuf->Start();
+        const uint16_t len = msgBuf->DataLength();
+
+        VerifyOrExit(len == 4, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+        PauseTimeMillis = LittleEndian::Read32(p);
+        mReliableMessageContext.HandleThrottleFlow(PauseTimeMillis);
+
+        ExitNow(err = CHIP_NO_ERROR);
+    }
+    // Return and not pass this to Application if Common::Null Msg Type
+    else if ((protocolId == Protocols::kProtocol_Protocol_Common) && (messageType == Protocols::Common::kMsgType_Null))
+    {
+        ExitNow(err = CHIP_NO_ERROR);
     }
     else
     {
-        DefaultOnMessageReceived(this, packetHeader, protocolId, messageType, std::move(msgBuf));
+        // Since we got the response, cancel the response timer.
+        CancelResponseTimer();
+
+        // If the context was expecting a response to a previously sent message, this message
+        // is implicitly that response.
+        SetResponseExpected(false);
+
+        if (mDelegate != nullptr)
+        {
+            mDelegate->OnMessageReceived(this, packetHeader, protocolId, messageType, std::move(msgBuf));
+        }
+        else
+        {
+            DefaultOnMessageReceived(this, packetHeader, protocolId, messageType, std::move(msgBuf));
+        }
```

Comment:
Again, extract these into `ReliableMessageContext`

Reply:
Same, this part is for another response mechanism operating on the protocol level, not for CRMP.